libraries/AP_Mount/AP_Mount_Siyi.cpp

```diff
@@ -0,0 +1,614 @@
+
+#define AP_MOUNT_SIYI_DEBUG 0
+#define debug(fmt, args ...) do { if (AP_MOUNT_SIYI_DEBUG) { GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Siyi: " fmt, ## args); } } while (0)
+
+AP_Mount_Siyi::AP_Mount_Siyi(AP_Mount &frontend, AP_Mount_Params &params, uint8_t instance) :
+    AP_Mount_Backend(frontend, params, instance)
+{}
+
+// init - performs any required initialisation for this instance
+void AP_Mount_Siyi::init()
+{
+    const AP_SerialManager& serial_manager = AP::serialmanager();
+
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_SToRM32, 0);
+    if (_uart != nullptr) {
+        _initialised = true;
+        set_mode((enum MAV_MOUNT_MODE)_params.default_mode.get());
+    }
+
+}
+
+// update mount position - should be called periodically
+void AP_Mount_Siyi::update()
+{
+    // exit immediately if not initialised
+    if (!_initialised) {
+        return;
+    }
+
+    // reading incoming packets from gimbal
+    read_incoming_packets();
+
+    // request firmware version at 1hz
+    uint32_t now_ms = AP_HAL::millis();
+    if (!_got_firmware_version) {
+        if ((now_ms - _last_send_ms) >= 1000) {
+            request_firmware_version();
+            _last_send_ms = now_ms;
+        }
+        return;
+    }
+
+    // request attitude at 10hz
+    if ((now_ms - _last_req_current_angle_rad_ms) >= 50) {
+        request_gimbal_attitude();
+        _last_req_current_angle_rad_ms = now_ms;
+    }
+
+    // update based on mount mode
+    switch (get_mode()) {
+        // move mount to a "retracted" position.  To-Do: remove support and replace with a relaxed mode?
+        case MAV_MOUNT_MODE_RETRACT: {
+            const Vector3f &angle_bf_target = _params.retract_angles.get();
+            send_target_angles(ToRad(angle_bf_target.y), ToRad(angle_bf_target.z), false);
+            break;
+        }
+
+        // move mount to a neutral position, typically pointing forward
+        case MAV_MOUNT_MODE_NEUTRAL: {
+            const Vector3f &angle_bf_target = _params.neutral_angles.get();
+            send_target_angles(ToRad(angle_bf_target.y), ToRad(angle_bf_target.z), false);
+            break;
+        }
+
+        // point to the angles given by a mavlink message
+        case MAV_MOUNT_MODE_MAVLINK_TARGETING:
+            switch (mavt_target.target_type) {
+            case MountTargetType::ANGLE:
+                send_target_angles(mavt_target.angle_rad.pitch, mavt_target.angle_rad.yaw, mavt_target.angle_rad.yaw_is_ef);
+                break;
+            case MountTargetType::RATE:
+                send_target_rates(mavt_target.rate_rads.pitch, mavt_target.rate_rads.yaw, mavt_target.rate_rads.yaw_is_ef);
+                break;
+            }
+            break;
+
+        // RC radio manual angle control, but with stabilization from the AHRS
+        case MAV_MOUNT_MODE_RC_TARGETING: {
+            // update targets using pilot's rc inputs
+            MountTarget rc_target {};
```

Comment:
I find it a little odd that we store both rates and angles in the same object.

Reply:
Txs for the review.

I've done it this way in all the other drivers as well as part of that major refactoring I did so it is at least consistent.

I do agree that it is in slightly bad taste to mix angles and rates like this but even if I created two structures they would be identical.. it would just be their names that were different.

```
    // structure for a single angle or rate target
    struct MountTarget {
        float roll;
        float pitch;
        float yaw;
        bool yaw_is_ef;
    };
```

Reply:
> I've done it this way in all the other drivers as well as part of that major refactoring I did so it is at least consistent.

OK.

> I do agree that it is in slightly bad taste to mix angles and rates like this but even if I created two structures they would be identical.. it would just be their names that were different.

Sure, that's OK.  It doesn't cost us extra to have two structures with similar field names.  Using two structures *does* give us type-safety, however.

Definitely not a cleanup to do on this PR.